accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,502 @@
+    {
+        /// <summary>
+        /// Protected constructor to create the instrument with the common properties.
+        /// </summary>
+        protected Instrument(Meter meter, string name, string? description, string? unit) :
+                        base(meter, name, description, unit) { throw null; }
+
+        /// <summary>
+        /// Record measurement overloads allowing passing different numbers of tags.
+        /// </summary>
+
+        protected void RecordMeasurement(T val) { throw null; }
+
+        protected void RecordMeasurement(
+                            T val,
+                            KeyValuePair<string, object> tag1) { throw null; }
+
+        protected void RecordMeasurement(
+                            T val,
+                            KeyValuePair<string, object> tag1,
+                            KeyValuePair<string, object> tag2) { throw null; }
+
+        protected void RecordMeasurement(
+                            T val,
+                            KeyValuePair<string, object> tag1,
+                            KeyValuePair<string, object> tag2,
+                            KeyValuePair<string, object> tag3) { throw null; }
+
+        protected void RecordMeasurement(
+                            T val,
+                            ReadOnlySpan<KeyValuePair<string, object>> tags) { throw null; }
+    }
+}
+```
+
+### Observable Instrument Base class
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// ObservableInstrument<T> is the base class from which all observable instruments will inherit from.
+    /// Mainly It'll support the CLS compliant numerical types
+    /// </summary>
+    public abstract class ObservableInstrument<T> : Instrument where T : unmanaged
+    {
+        /// <summary>
+        /// Protected constructor to create the instrument with the common properties.
+        /// </summary>
+        protected ObservableInstrument(
+                    Meter meter,
+                    string name,
+                    string? description,
+                    string? unit) : base(meter, name, description, unit) { throw null; }
+
+        /// <summary>
+        /// Observe will report the non-additive values of the measurements.
+        /// </summary>
+        protected abstract IEnumerable<Measurement<T>> Observe();
+    }
+}
+```
+
+### Measurement class
+
+```csharp
+
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// A helper class used by the Observable instruments Observe method to repro the measurement values
+    /// with the associated tags.
+    /// </summary>
+    public struct Measurement<T> where T : unmanaged
+    {
+        /// <summary>
+        /// Construct the Measurement using the value and the list of tags.
+        /// We'll always copy the input list as this is not perf hot path.
+        /// </summary>
+        public Measurement(T value, IEnumerable<KeyValuePair<string, object>> tags) { throw null; }
```

Comment:
Do `object` here need to be nullable as `object?`

Reply:
Should we allow null object?

Reply:
yes. we already allow that in tags used in tracing.

Reply:
I found I already correctly annotated it in `public ReadOnlySpan<KeyValuePair<string, object?>> Tags { get { throw null;  } }` but missed the annotations in the other places. Thanks for the catch.